### ISSUE TITLE
Setting DEBUG = False in Wikiconfig.py has no effect with Built-In Server

### ISSUE THREAD

**Issue by ThomasWaldmann** [moinwiki/moin#546]
**[Original report](https://bitbucket.org/thomaswaldmann/moin-2.0/issue/546) by RogerHaase (Bitbucket: [RogerHaase](https://bitbucket.org/RogerHaase), GitHub: [RogerHaase](https://github.com/RogerHaase)).**

----------------------------------------

Near the bottom of wikiconfig.py there is:

    # DEBUG = False # use True for development only, not for public sites!

which is further explained in http://flask.pocoo.org/docs/0.10/errorhandling/#debugging-application-errors and http://werkzeug.pocoo.org/docs/0.11/debug/#module-werkzeug.debug 

Uncommenting the above line has no effect - the werkzeug debugger is still active when running the built-in server. 

In flask_script/command.py, there is this code setting the debugger on:

    class Server(Command):
        """
        Runs the Flask development server i.e. app.run()

        :param host: server host
        :param port: server port
        :param use_debugger: if False, will no longer use Werkzeug debugger.
                             This can be overriden in the command line
                             by passing the **-d** flag.
        :param use_reloader: if False, will no longer use auto-reloader.
                             This can be overriden in the command line by
                             passing the **-r** flag.
        :param threaded: should the process handle each request in a separate
                         thread?
        :param processes: number of processes to spawn
        :param passthrough_errors: disable the error catching. This means that the server will die on errors but it can be useful to hook debuggers in (pdb etc.)
        :param options: :func:`werkzeug.run_simple` options.
        """

        help = description = 'Runs the Flask development server i.e. app.run()'

        def __init__(self, host='127.0.0.1', port=5000, use_debugger=True,
                     use_reloader=True, threaded=False, processes=1,
                     passthrough_errors=False, **options):

            self.port = port
            self.host = host
            self.use_debugger = use_debugger
            self.use_reloader = use_reloader
            self.server_options = options
            self.threaded = threaded
            self.processes = processes
            self.passthrough_errors = passthrough_errors

Adding:

    USE_DEBUGGER = False
    USE_RELOADER = False

to wikiconfig.py has no effect.

There appears to be no code to override the above defaults using configuration variables. At the time the `__init__.py` code is executed, wikiconfig.py has not yet been processed.

The above comments regarding the -d and -r flags are unclear. Doing

    ./m run -d -r

results in turning off both the debugger and auto reloader.

Per the werkzeug docs:

    The interactive debugger however does not work in forking environments which makes it nearly impossible to use on production servers.

So this problem effects only wiki admins using the built-in server in production and maybe some other unidentified servers.

-- snip --

Using the Apache server, there is a small difference in behavior after a traceback dependent upon the DEBUG = True/False setting.

When a traceback occurs with DEBUG = True:

Browser displays: 

    Internal Server Error
    The server encountered an internal error or misconfiguration and was unable to complete your request.
    Please contact the server administrator, xxxxx@yahoo.com and inform them of the time the error occurred, and anything you might have done that may have caused the error.
    More information about this error may be available in the server error log

Log Traceback:

    [Sun Jan 10 13:21:04 2016] [error] Traceback (most recent call last):
    [Sun Jan 10 13:21:04 2016] [error]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1817, in wsgi_app
    [Sun Jan 10 13:21:04 2016] [error]     response = self.full_dispatch_request()
    [Sun Jan 10 13:21:04 2016] [error]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1477, in full_dispatch_request
    [Sun Jan 10 13:21:04 2016] [error]     rv = self.handle_user_exception(e)
    [Sun Jan 10 13:21:04 2016] [error]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1381, in handle_user_exception
    [Sun Jan 10 13:21:04 2016] [error]     reraise(exc_type, exc_value, tb)
    [Sun Jan 10 13:21:04 2016] [error]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1475, in full_dispatch_request
    [Sun Jan 10 13:21:04 2016] [error]     rv = self.dispatch_request()
    [Sun Jan 10 13:21:04 2016] [error]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1461, in dispatch_request
    [Sun Jan 10 13:21:04 2016] [error]     return self.view_functions[rule.endpoint](**req.view_args)
    [Sun Jan 10 13:21:04 2016] [error]   File "C:\\home\\web\\moin2\\MoinMoin\\apps\\frontend\\views.py", line 471, in show_item
    [Sun Jan 10 13:21:04 2016] [error]     assert False
    [Sun Jan 10 13:21:04 2016] [error] AssertionError
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1] mod_wsgi (pid=472): Exception occurred processing WSGI script 'C:/home/web/moin2/moin2.wsgi'., referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1] Traceback (most recent call last):, referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1836, in __call__, referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1]     return self.wsgi_app(environ, start_response), referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1820, in wsgi_app, referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1]     response = self.make_response(self.handle_exception(e)), referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1403, in handle_exception, referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1]     reraise(exc_type, exc_value, tb), referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1817, in wsgi_app, referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1]     response = self.full_dispatch_request(), referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1477, in full_dispatch_request, referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1]     rv = self.handle_user_exception(e), referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1381, in handle_user_exception, referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1]     reraise(exc_type, exc_value, tb), referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1475, in full_dispatch_request, referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1]     rv = self.dispatch_request(), referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1461, in dispatch_request, referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1]     return self.view_functions[rule.endpoint](**req.view_args), referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1]   File "C:\\home\\web\\moin2\\MoinMoin\\apps\\frontend\\views.py", line 471, in show_item, referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1]     assert False, referer: http://packrat.digitalrockart.org/moin2/rst
    [Sun Jan 10 13:21:04 2016] [error] [client 127.0.0.1] AssertionError, referer: http://packrat.digitalrockart.org/moin2/rst

With DEBUG = False:

Browser displays:

    Internal Server Error
    The server encountered an internal error and was unable to complete your request. Either the server is overloaded or there is an error in the application.

Error log traceback:

    [Sun Jan 10 13:15:18 2016] [error] Traceback (most recent call last):
    [Sun Jan 10 13:15:18 2016] [error]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1817, in wsgi_app
    [Sun Jan 10 13:15:18 2016] [error]     response = self.full_dispatch_request()
    [Sun Jan 10 13:15:18 2016] [error]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1477, in full_dispatch_request
    [Sun Jan 10 13:15:18 2016] [error]     rv = self.handle_user_exception(e)
    [Sun Jan 10 13:15:18 2016] [error]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1381, in handle_user_exception
    [Sun Jan 10 13:15:18 2016] [error]     reraise(exc_type, exc_value, tb)
    [Sun Jan 10 13:15:18 2016] [error]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1475, in full_dispatch_request
    [Sun Jan 10 13:15:18 2016] [error]     rv = self.dispatch_request()
    [Sun Jan 10 13:15:18 2016] [error]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1461, in dispatch_request
    [Sun Jan 10 13:15:18 2016] [error]     return self.view_functions[rule.endpoint](**req.view_args)
    [Sun Jan 10 13:15:18 2016] [error]   File "C:\\home\\web\\moin2\\MoinMoin\\apps\\frontend\\views.py", line 471, in show_item
    [Sun Jan 10 13:15:18 2016] [error]     assert False
    [Sun Jan 10 13:15:18 2016] [error] AssertionError
    [Sun Jan 10 13:15:18 2016] [error] 2016-01-10 13:15:18,403 ERROR MoinMoin:1423 Exception on /Home [GET]
    [Sun Jan 10 13:15:18 2016] [error] Traceback (most recent call last):
    [Sun Jan 10 13:15:18 2016] [error]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1817, in wsgi_app
    [Sun Jan 10 13:15:18 2016] [error]     response = self.full_dispatch_request()
    [Sun Jan 10 13:15:18 2016] [error]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1477, in full_dispatch_request
    [Sun Jan 10 13:15:18 2016] [error]     rv = self.handle_user_exception(e)
    [Sun Jan 10 13:15:18 2016] [error]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1381, in handle_user_exception
    [Sun Jan 10 13:15:18 2016] [error]     reraise(exc_type, exc_value, tb)
    [Sun Jan 10 13:15:18 2016] [error]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1475, in full_dispatch_request
    [Sun Jan 10 13:15:18 2016] [error]     rv = self.dispatch_request()
    [Sun Jan 10 13:15:18 2016] [error]   File "C:\\home\\web\\moin2-venv-python\\Lib\\site-packages\\flask\\app.py", line 1461, in dispatch_request
    [Sun Jan 10 13:15:18 2016] [error]     return self.view_functions[rule.endpoint](**req.view_args)
    [Sun Jan 10 13:15:18 2016] [error]   File "C:\\home\\web\\moin2\\MoinMoin\\apps\\frontend\\views.py", line 471, in show_item
    [Sun Jan 10 13:15:18 2016] [error]     assert False
    [Sun Jan 10 13:15:18 2016] [error] AssertionError

----
In conclusion, that "DEBUG = False" line in wikiconfig.py is misleading and should be clarified. Moin docs should be updated to say if you want to run built-in server in production, use

    ./m run -d -r

(or do it the long way, activate the env and do "moin -d -r")
